### PR TITLE
Add support for .NET Core 3.0.

### PR DESCRIPTION
**Pull request by laura-rodriguez**
* Recreate all .NET Core projects (libraries and test projects) using .NET Core 3.0 so we make sure we are using the latest .NET Core project template. All the existent classes in the current Okta.AspNetCore library will be ported to the new version, and we will modify only those that need any adjustments due to breaking changes in the .NET Core framework. 

* We will still reuse `Okta.AspNet.Abstractions` as we do in the current version. This project has all the common classes used by all the Okta ASP.NET libraries.

* Release a major version (3.0.0 to be consistent with the .NET Core version).

* Since we will still support ASP.NET Core 2.x, create a branch for the current version. Any changes needed in this version will be merged into this branch. Master will now have the Okta.AspNetCore 3.X version.

* Fix StyleCop warnings and remove suppressions.

### CODE REVIEW

**Comment by chuanbozhang-okta** [comment]
Is there a reason we're not doing .net core 3.1 which is LTS (long term support) release? 
.net core 3.0 will be end of support in less than 4 months while 3.1 will be supported for 3 years. See https://dotnet.microsoft.com/platform/support/policy/dotnet-core for official support policy

Version | Original Release Date | Latest Patch Version | Patch Release Date | Support Level | End of Support
-- | -- | -- | -- | -- | --
.NET Core 3.1 | December 3, 2019 | 3.1.0 | 3.1.0 | LTS |  
.NET Core 3.0 | September 23, 2019 | 3.0.1 | November 19, 2019 | Maintenance | March 3, 2020

**Comment by laura-rodriguez** [comment]
Hi @chuanbozhang-okta ,

Thanks for asking. The reason is that we started with this task before 3.1 was released. After we merge this into master, we might need to iterate this once more to upgrade to 3.1.

File: Okta.AspNetCore.Mvc.IntegrationTest/Controllers/HomeController.cs [hunk 1]
```diff
@@ -1,38 +0,0 @@
-﻿// <copyright file="HomeController.cs" company="Okta, Inc">
```
Comment (bradhaverstein-okta) [inline Okta.AspNetCore.Mvc.IntegrationTest/Controllers/HomeController.cs]: I pinged Laura offline to ask why we were removing the controllers and static files from `Okta.AspNetCore.Mvc.IntegrationTest`. It sounds like these aren't actually needed by the integration tests since the tests are only concerned with startup logic.

Just a nitpick but I'd recommend adding a line to the PR description to explain the motivation for removing those files. That seems like a change which isn't directly related to the .NET core version upgrade so its good to note why it is happening.

File: Okta.AspNetCore.Test/Okta.AspNetCore.Test.csproj [hunk 2]
```diff
@@ -1,26 +1,22 @@
...
-    <ProjectReference Include="..\Okta.AspNetCore\Okta.AspNetCore.csproj" />
-    <PackageReference Include="AsyncUsageAnalyzers" Version="1.0.0-alpha003" />
-    <PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta007" />
-    <AdditionalFiles Include="..\stylecop.json" />
+    <PackageReference Include="FluentAssertions" Version="5.9.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
+    <PackageReference Include="NSubstitute" Version="4.2.1" />
+    <PackageReference Include="xunit" Version="2.4.0" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.0" />
+    <PackageReference Include="coverlet.collector" Version="1.0.1" />
```
Comment (bradhaverstein-okta) [inline Okta.AspNetCore.Test/Okta.AspNetCore.Test.csproj]: Were we using `coverlet.collector` before the .NET core version upgrade? If we are introducing a dependency on `coverlet.collector` I think it would be good to explain why in the PR description.

If we were using it before and this is just a new reference required for .NET core 3 then the current PR description is fine.

File: Okta.AspNetCore.WebApi.IntegrationTest/Okta.AspNetCore.WebApi.IntegrationTest.csproj [hunk 3]
```diff
@@ -1,45 +1,27 @@
...
-  </ItemGroup>
-
-  <ItemGroup>
-    <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="2.0.2" />
-  </ItemGroup>
-
-  <ItemGroup>
-    <ProjectReference Include="..\Okta.AspNet.Abstractions\Okta.AspNet.Abstractions.csproj" />
     <ProjectReference Include="..\Okta.AspNetCore\Okta.AspNetCore.csproj" />
-    <AdditionalFiles Include="..\stylecop.json" />
```
Comment (bradhaverstein-okta) [inline Okta.AspNetCore.WebApi.IntegrationTest/Okta.AspNetCore.WebApi.IntegrationTest.csproj]: nit: Are we intentionally removing this `<AdditionalFiles >` element or is this something we forgot to add back when recreating the project?

File: Okta.AspNetCore.WebApi.IntegrationTest/appsettings.json [hunk 4]
```diff
@@ -1,15 +1,10 @@
...
-        "Default": "Warning"
-      }
+    "LogLevel": {
+      "Default": "Information",
+      "Microsoft": "Warning",
+      "Microsoft.Hosting.Lifetime": "Information"
     }
-  }
+  },
+  "AllowedHosts": "*"
```
Comment (bradhaverstein-okta) [inline Okta.AspNetCore.WebApi.IntegrationTest/appsettings.json]: I'm not very familiar with the `AllowedHosts` config setting, but I was [reading here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-3.1#endpoint-configuration) it the docs made it seem like omitting it altogether would prevent the host filtering middleware from even running. I was wondering if this would have the same effect as specifying `*` for the host filter.

Would it be better to omit this setting or is it needed for some reason?

File: Okta.AspNetCore/GlobalSuppressions.cs [hunk 5]
```diff
@@ -1,7 +0,0 @@
-﻿// <copyright file="GlobalSuppressions.cs" company="Okta, Inc">
-// Copyright (c) 2018-present Okta, Inc. All rights reserved.
-// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.
-// </copyright>
-
-[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.DocumentationRules", "SA1600:Elements should be documented", Justification = "Reviewed")]
-[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.DocumentationRules", "SA0001XmlCommentAnalysisDisabled", Justification = "Reviewed")]
```
Comment (bradhaverstein-okta) [inline Okta.AspNetCore/GlobalSuppressions.cs]: It seems like we are removing these suppressions by adding documentation to all the public members, is that right? If that's the case we should probably note that in the PR description, since this is not directly related to upgrading to .NET core 3.0.

**Comment by laura-rodriguez** [comment]
> Looks pretty good to me.
> 
> One overall comment I have is that I wonder whether we could produce one library that supports both .NET Core 3.0 instead of having a separate .NET Core 3.0 branch. The [Microsoft docs](https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.1&tabs=visual-studio#migrate-libraries-via-multi-targeting) seem to indicate this is possible by using multi-targeting. But I'll leave that strategic decision up to the SDK team.

We decided not to follow the multi-targeting approach to avoid using preprocessor ifs when we have to deal with features that are not available in older .net core versions. For example, 3.0 added PKCE out of the box, so if we want to be consistent across .NET versions, we will have to make a custom PKCE implementation for older .Net core versions and the code would look like this:
```
oidcOptions.UseTokenLifetime = false;
#if NETCOREAPP3_0 
            oidcOptions.UsePkce = true;
#elseif NETCOREAPP2_0
...
#endif
```
This will be a pain to maintain in the future when more new features are added to the newest .net core version.

File: Okta.AspNetCore/GlobalSuppressions.cs (hunk 5 above)
Comment (laura-rodriguez) [reply Okta.AspNetCore/GlobalSuppressions.cs]: You're right. I removed those suppressions since I fixed those warnings.

File: Okta.AspNetCore.WebApi.IntegrationTest/Okta.AspNetCore.WebApi.IntegrationTest.csproj (hunk 3 above)
Comment (laura-rodriguez) [reply Okta.AspNetCore.WebApi.IntegrationTest/Okta.AspNetCore.WebApi.IntegrationTest.csproj]: This is intentional. We now use StyleCop only for the libraries, and not in tests.

File: Okta.AspNetCore.Test/Okta.AspNetCore.Test.csproj (hunk 2 above)
Comment (laura-rodriguez) [reply Okta.AspNetCore.Test/Okta.AspNetCore.Test.csproj]: Nice catch! I believe this came with the new test project template. I'll double check.

File: Okta.AspNetCore.WebApi.IntegrationTest/appsettings.json (hunk 4 above)
Comment (laura-rodriguez) [reply Okta.AspNetCore.WebApi.IntegrationTest/appsettings.json]: I believe we can omit this setting in this case. I'll double check.
